Documentation/design-docs/tailcalls-with-helpers.md

```diff
@@ -0,0 +1,266 @@
+* TailCallHelperStack - a per thread stack of helper entries that is used to determine whether a tail call is chained or not. Its entries are allocated as local variables in CallTarget thunks. Each entry contains:
+  * Stack pointer captured right before a call to a tail call target
+  * ChainCall flag indicating whether the CallTarget thunk should return after the call to the tail call target or whether it should execute its epilog and jump to TailCallHelper instead. The latter is used by the TailCallHelper to remove the stack frame of the CallTarget before making a tail call from a tail called function. 
+  * Pointer to the next entry on the stack.
+
+### Tail call sequence
+* The caller calls the StoreArguments thunk corresponding to the callee to store the arguments, their GC descriptors and the corresponding CallTarget thunk address in a thread local storage.
+* The caller executes its epilog, restoring stack pointer and callee saved registers to their values when the caller was entered.
+* The caller jumps to the TailCallHelper. This function performs the following operations:
+  * Get the topmost TailCallHelperStack entry for the current thread.
+  * Check if the previous stack frame is a CallTarget thunk frame by comparing the stack pointer value stored in the TailCallHelperStack entry to the current CFA (call frame address). If it matches, it means that the previous stack frame belongs to a CallTarget thunk and so the tail call caller was also tail called.
+  * If the previous frame was a CallTarget thunk, its stack frame needs to be removed to ensure that the stack will not grow when tail calls are chained. Set ChainCall flag in the TailCallHelperStack entry and return. That returns control to the CallTarget thunk. It checks the ChainCall flag and since it set, it executes its epilog and jumps to the TailCallHelper again.
+  * If the previous frame was not a CallTarget thunk, get the address of the CallTarget thunk of the tailcall target from the arguments buffer and jump to it.
+* The CallTarget thunk function then does the following operation:
+  * Create local instance of TailCallHelperStack entry and store the current stack pointer value in it.
+  * Push the entry to the TailCallHelperStack of the current thread.
+  * Get the arguments buffer from the thread local storage, extract the arguments, release the buffer and call the target function. The frame of the CallTarget thunk ensures that the arguments of the target are GC protected until the target function returns or tail calls to another function.
+  * Pop the TailCallHelperStack entry from the TailCallHelperStack of the current thread.
+  * Check the ChainCall flag in the TailCallHelperStack entry. If it is set, run epilog and jump to the TailCallHelper.
+  * If the ChainCall flag is clear, it means that the last function in the tail call chain has returned. So return the return value of the target function.
```

Comment:
On further thinking, it looks it will work for single-level tail calls but not for chained tail calls.

Let us consider an example tail call chain

A() -> calls B() --> tail call C() 

Assume further B(), C() return a gc-type return value.

Here is how the call stack looks while executing inside C().

A() -> CallTarget of C --> C()

Say C() in turn tail call into D(), which also return a gc-type return value.

C() would call StoreArgument thunk to store args for D in TLS and jumps to TailCallHelper after executing its epilog.  Essentially call stack looks as

A() --> CallTarget of C --> TailCallHelper()

TailCallHelper() finds that previous stack frame is CallTarget() and hence marks it as chained and returns to CallTarget of C().  This will appear to CallTarget as if C() has returned, but return register do not contain a valid gc-ref.  
Say if GC kicks in immediate after returning to CallTarget of C.  As per gc-info of CallTarget of C, return reg (say RAX) contains a valid gc-ref but it does not. 

Q1: Won't it lead to a crash when GC tries to use in invalid gc-ref in rax and mark it? Is my understanding right here?

Call stack after TailCalHelper() returns is

A() --> CallTarget of C

CallTarget on finding ChainCall Flag, executes its epilog and jumps to TailCallHelper.  What is the return value of TailCallHelper()?  Since it is common for all tail call targets, I am assuming to be void.  

Q2: Can GC kick-in while executing inside TailCallHelper()?

Call stack at this point is

A() --> TailCallHelper()

TailCallHelper() in turn now jumps to CallTarget of D().  The call stack now is

A() --> CallTarget of D --> D()



---
In reply to: [92481487](https://github.com/dotnet/coreclr/pull/8596#discussion_r92481487) [](ancestors = 92481487,92259054)